### PR TITLE
[WIP]

### PR DESCRIPTION
**Pull request by antmat**
New logging service.
Work is DEEPLY in progress, do not merge.
UPD.
The goal of this service is to provide logging capability with more granular control. It will help us to avoid unnecessary logs and understand what's happening in case of some problems by increasing the amount of debug information. I'll describe main usage pattern.
- logging service initial dispatch has only one method "get", which returns named logger. After that there is a transition to dispatch with emit method, which is used to emit logs. All log messages will have an attribute with name of the logger.
- logging control service provides interface to manage filters for named loggers. There will be a reserved name to add filters to core logger in cocaine. There is a possibility to extract core logger control to separate service, but it will duplicate API of this service, so I don't like it.
- Filter itself can be represented as dynamic structure with 2 or 3 elements.
  - operator. One of
    - "e" - attribute exists
    - "!e" - attribute do not exist
    - "==" - attribute value equal to
    - "!=" - attribute value not equal to
    - ">" - attribute value greater than
    - ">=" - attribute value greater or equal than
    - "<" - attribute value less than
    - "<=" - attribute value less or equal than
    - "||" - apply OR operation to following filters
    - "&&" - apply AND operation to following filters
    - "xor" - apply XOR operation to following filters
  - first operand:
    - In case of  "e", "!e", "==", "!=", ">", ">=", "<", "<=" - attribute name
    - In case of "||", "&&", "xor" - recursive description of first filter
  - second operand:
    - In case of  "e", "!e" - missing
    - In case of "==", "!=", ">", ">=", "<", "<=" - attribute value
    - In case of "||", "&&", "xor" - recursive description of second filter

Development plan:
- Refactor unicorn to get named configuration of zookeper via repository(as it is done in f.e. storage)
- Implement logging service itself with transition to named logger
- Implement bind of named logger to control class to fetch filter information
- Implement control service with described api.
  - filter_t msgpack trait
  - filter_t information class (id, timer is cluster wide etc)
- Implement bind of control service to control class which actually stores filtering information
- Implement control class

### CODE REVIEW

File: logging/include/cocaine/idl/logging.hpp [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include "cocaine/filter.hpp"
```
Comment (3Hren) [inline logging/include/cocaine/idl/logging.hpp]: Replace with `#include "cocaine/<plugin_name>/filter.hpp"` to avoid conflicts with core includes.

**Comment by kobolog** [comment]
Arguably, Logging Service should be a part of the Core.

File: logging/include/cocaine/logging/control.hpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    filter_t::id_type
+    add_filter(std::string logger_name, std::shared_ptr<filter_t> filter);
+
+    bool
+    remove_filter(filter_t::id_type id);
+private:
+    std::atomic<filter_t::id_type>
+    zookeeper_api_t zk;
+    synchronized<std::map<std::string, std::shared_ptr<metafilter_t>>> metafilters;
+    synchronized<std::map<std::string, std::shared_ptr<metafilter_t>>> metafilters;
```
Comment (3Hren) [inline logging/include/cocaine/logging/control.hpp]: Have you ever tried to compile this?

File: logging/include/cocaine/logging/filter.hpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include <stdint.h>
```
Comment (3Hren) [inline logging/include/cocaine/logging/filter.hpp]: Use c++ aliases (`<cstdint>`).

File: logging/include/cocaine/logging/metafilter.hpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        ptr_type
+        get(const std::string& name) = 0;
+    };
+
+    template<class T>
+    struct default_factory: public factory_type {
+        virtual
+        ptr_type
+        get(const std::string& name) {
+            return instances.apply([&](std::map<std::string, std::weak_ptr<storage_t>>& instances) {
```
Comment (3Hren) [inline logging/include/cocaine/logging/metafilter.hpp]: This is already looks like a boilerplate. I remember at least 3 different parts of code where such construction occurs. Maybe decompose it into something like `cached_factory`?

File: logging/include/cocaine/traits/filter.hpp [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace cocaine { namespace io {
+
+template<>
+struct type_traits<logging::filter_t> {
+    template<class Stream>
+    static inline
+    void
+    pack(msgpack::packer<Stream>& target, const logging::filter_t& source) {
+    }
```
Comment (3Hren) [inline logging/include/cocaine/traits/filter.hpp]: Why empty?

File: logging/src/logging.cpp [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+namespace cocaine { namespace service {
+
+const io::basic_dispatch_t&
+logging_t::prototype() {
+    return *this;
+}
+
+logging_t::logging_t(context_t& context, asio::io_service& asio, const std::string& name, const dynamic_t& args)
+{
+    typedef typename io::basic_slot<io::base_log::get>::dispatch_type get_dispatch_type;
```
Comment (3Hren) [inline logging/src/logging.cpp]: It's local typedefs, just `tuple_type` etc.

File: logging/src/logging.cpp [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        record.insert(cocaine::logging::keyword::source() = std::move(source));
+        record.message(std::move(message));
+
+        wrapper->push(std::move(record));
+    }
+}
+
+const io::basic_dispatch_t&
+logging_control_t::prototype() const {
+
```
Comment (3Hren) [inline logging/src/logging.cpp]: GJ!

File: logging/src/logging.cpp [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+logging_control_t::logging_control_t(context_t& context, asio::io_service& asio, const std::string& name, const dynamic_t& args) {
+
```
Comment (3Hren) [inline logging/src/logging.cpp]: Mark with `BOOST_ASSERT(false)` or similar construction to help us see is this part really unimplemented or it is empty just as planned.

**Comment by 3Hren** [comment]
This logging service requires `unicorn` to work properly which cannot be guaranteed to be loaded in runtime. So I've suggested to make this as a `logging::v2` plugin.

File: logging/include/cocaine/logging/metafilter.hpp (hunk 4 above)
Comment (kobolog) [inline logging/include/cocaine/logging/metafilter.hpp]: Yeah, I've been thinking about it for a long time. We essentially have two types of factories now – the one that creates a new instance every call and returns `std::unique_ptr<T>` and the one that keeps a map of `std::weak_ptr<T>` and returns a cached instance if it's still in use somewhere.

**Comment by kobolog** [comment]
Well in this case it's not really a v2, it's like the Core one is a Basic Logging Service and this one is like an Advanced one or something like that.

**Comment by 3Hren** [comment]
@antmat write please what's going on here. I want to see:

 - Why is it required.
 - How you've implemented this.
 - What's next (desirable with checkboxes).
 - Usage example code from any framework you like.

File: logging/include/cocaine/idl/logging.hpp [hunk 9]
```diff
@@ -0,0 +1,218 @@
+/*
```
Comment (3Hren) [inline logging/include/cocaine/idl/logging.hpp]: This header **will** conflict with the similar in a **cocaine-core**.

File: logging/include/cocaine/logging/control.hpp (hunk 2 above)
Comment (antmat) [inline logging/include/cocaine/logging/control.hpp]: nope. It's just a skeleton.

File: logging/include/cocaine/traits/filter.hpp (hunk 5 above)
Comment (antmat) [inline logging/include/cocaine/traits/filter.hpp]: It's a stub. Anyway compiler will remind me with "unused variables warning"

File: logging/src/logging.cpp (hunk 6 above)
Comment (antmat) [inline logging/src/logging.cpp]: In case we have more methods they could conflict.

File: logging/src/logging.cpp (hunk 7 above)
Comment (antmat) [inline logging/src/logging.cpp]: THX!

File: logging/src/logging.cpp (hunk 8 above)
Comment (antmat) [inline logging/src/logging.cpp]: ok

**Pull request closed without merging**